src/vec_int128_ppc.h

```diff
@@ -1853,6 +1963,157 @@ vec_cmul10cuq (vui128_t *cout, vui128_t a)
+ *  |power8   | 18-60 | 1/cycle  |
+ *  |power9   | 20-45 | 1/cycle  |
+ *
+ *  @param vra the dividend as a vector treated as a unsigned __int128.
+ *  @return the quotient as vector unsigned __int128.
+ */
+static inline vi128_t
+vec_divsq_10e31 (vi128_t vra)
+{
+  const vui128_t zero = (vui128_t) vec_splats ((int) 0);
+  /* ten32  = +100000000000000000000000000000000UQ  */
+  const vui128_t ten31 = (vui128_t)
+	  { (__int128) 1000000000000000UL * (__int128) 10000000000000000UL };
+  /* Magic numbers for multiplicative inverse to divide by 10**31
+   are 4804950418589725908363185682083061167, Corrective add,
+   and shift right 107 bits.  */
+  const vui128_t mul_invs_ten31 = (vui128_t) CONST_VINT128_DW(
+      0x039d66589687f9e9UL, 0x01d59f290ee19dafUL);
+  const int shift_ten31 = 103;
+  vui128_t result, t, q, uvra;
+  vb128_t negbool;
+
+  negbool = vec_setb_sq (vra);
+  uvra = (vui128_t) vec_sel ((vui32_t) vra,
+  			    (vui32_t) vec_subuqm (zero, (vui128_t) vra),
+  			    (vb32_t) negbool);
+
+  if (vec_cmpuq_all_ge (uvra, ten31))
+    {
+      q = vec_mulhuq (uvra, mul_invs_ten31);
+      // Need corrective add but don't want to deal with the carry.
+      t = vec_subuqm (uvra, q);
+      t = vec_srqi (t, 1);
+      t = vec_adduqm (t, q);
+      result = vec_srqi (t, (shift_ten31 - 1));
+      result = (vui128_t) vec_sel ((vui32_t) result,
+      			    (vui32_t) vec_subuqm (zero, (vui128_t) result),
+      			    (vb32_t) negbool);
+    }
+  else
+    result = (vui128_t) vec_splat_u32(0);
+
+  return (vi128_t) result;
+}
+
+/** \brief Vector Divide by const 10e31 Unsigned Quadword.
+ *
+ *  Compute the quotient of a 128 bit values vra / 10e31.
+ *
+ *  \note vec_divuq_10e31() and vec_moduq_10e31() can be used to
+ *  prepare for <B>Decimal Convert From Signed Quadword</B>
+ *  (See vec_bcdcfsq()), This guarantees that the conversion to
+ *  Vector BCD does not overflow and the 39-digit extended result is
+ *  obtained.
+ *
+ *  |processor|Latency|Throughput|
+ *  |--------:|:-----:|:---------|
+ *  |power8   |  8-48 | 1/cycle  |
+ *  |power9   |  9-31 | 1/cycle  |
+ *
+ *  @param vra the dividend as a vector treated as a unsigned __int128.
+ *  @return the quotient as vector unsigned __int128.
+ */
+static inline vui128_t
+vec_divuq_10e31 (vui128_t vra)
+{
+  /* ten32  = +100000000000000000000000000000000UQ  */
+  const vui128_t ten31 = (vui128_t)
+	  { (__int128) 1000000000000000UL * (__int128) 10000000000000000UL };
+  /* Magic numbers for multiplicative inverse to divide by 10**31
+   are 4804950418589725908363185682083061167, Corrective add,
+   and shift right 107 bits.  */
+  const vui128_t mul_invs_ten31 = (vui128_t) CONST_VINT128_DW(
+      0x039d66589687f9e9UL, 0x01d59f290ee19dafUL);
+  const int shift_ten31 = 103;
+  vui128_t result, t, q;
+
+  if (vec_cmpuq_all_ge (vra, ten31))
+    {
+      q = vec_mulhuq (vra, mul_invs_ten31);
+      // Need corrective add but don't want to deal with the carry.
```

Comment:
...seems like there should be a next sentence "So, ..."
This construction occurs in a few other places as well.

Reply:
ya this is based on one of the examples from Hackers delight page 229, The carry and the double quadword shift would be a pain at this point. 

so something like:
      // Need corrective add but don't want to deal with the carry.
      // The following is slightly longer but avoids the carry.
      t = vec_subuqm (vra, q);
      t = vec_srqi (t, 1);
      t = vec_adduqm (t, q);
      result = vec_srqi (t, (shift_ten31 - 1));

Reply:
Final wording

      // Need corrective add but want to avoid carry & double quad shift
      // The following avoids the carry and less instructions